File: client/gwt/src/main/java/fr/gael/dhus/gwt/services/UserService.java
/*
 * Data Hub Service (DHuS) - For Space data distribution.
 * Copyright (C) 2013,2014,2015 GAEL Systems
 *
 * This file is part of DHuS software sources.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package fr.gael.dhus.gwt.services;

import java.util.Date;
import java.util.HashMap;
import java.util.List;

import com.google.gwt.user.client.rpc.RemoteService;

import fr.gael.dhus.gwt.share.ProductData;
import fr.gael.dhus.gwt.share.SearchData;
import fr.gael.dhus.gwt.share.UserData;
import fr.gael.dhus.gwt.share.exceptions.UserServiceException;
import fr.gael.dhus.gwt.share.exceptions.UserServiceMailingException;
import fr.gael.dhus.gwt.share.exceptions.UserServiceNotExistingException;

public interface UserService extends RemoteService
{
   public List<UserData> getUsers (int start, int count, String filter)
      throws UserServiceException;

   public Integer count (String filter) throws UserServiceException;

   public void createUser (UserData userData) throws UserServiceException,
      UserServiceMailingException;

   public void createTmpUser (UserData userData) throws UserServiceException,
      UserServiceMailingException;

   public void updateUser (UserData userData) throws UserServiceException,
      UserServiceMailingException;

   public void deleteUser (Long id) throws UserServiceException,
      UserServiceMailingException;

   public UserData getUser (Long id) throws UserServiceException;

   public UserData getUserWithDataAccess (Long id) throws UserServiceException;

   public void updateDataAccess(UserData userData) throws UserServiceException;
   
   public void forgotPassword(UserData userData) throws UserServiceException,
      UserServiceMailingException, UserServiceNotExistingException;
   
   public void selfUpdateUser (UserData userData) throws UserServiceException,
      UserServiceMailingException;
   
   public void selfChangePassword(Long id, String oldPassword, String newPassword) throws UserServiceException,
      UserServiceMailingException;
   
   public void storeUserSearch (Long id, String search, String footprint, HashMap<String, String> advanced, String complete) throws UserServiceException;
   
   public void removeUserSearch (Long uId, Long sId) throws UserServiceException;
   
   public List<SearchData> getAllUserSearches (Long uId) throws UserServiceException;
   
   public int countUserSearches (Long uId) throws UserServiceException;
   
   public List<SearchData> scrollSearchesOfUser (int start, int count, Long uId)
            throws UserServiceException;
   
   public void clearSavedSearches (Long uId) throws UserServiceException;
   
   public List<ProductData> getUploadedProducts(int start, int count, Long uId) throws UserServiceException;
   public List<String> getUploadedProductsIdentifiers(int start, int count, Long uId) throws UserServiceException;
   public int countUploadedProducts (Long uId) throws UserServiceException;
   
   public void activateUserSearchNotification (Long sId, boolean notify) throws UserServiceException;
   
   public Date getNextScheduleSearch() throws UserServiceException;

   public List<UserData> getAllUsers (int start, int count, String filter)
      throws UserServiceException;

   public Integer countAll (String filter) throws UserServiceException;
   
   public Boolean checkUserCodeForPasswordReset(String code) throws UserServiceException;
   
   public void resetPassword(String code, String newPassword) throws UserServiceException,
   UserServiceMailingException;
   
   public boolean isDataPublic() throws UserServiceException;

   public List<UserData> getUsersForDataRight (int start, int count, String filter)
      throws UserServiceException;

   public Integer countForDataRight (String filter) throws UserServiceException;

   public UserData getPublicData() throws UserServiceException;
}

File: client/webclient/src/main/frontend/app/scripts/services/user-service.js
/* 
 * Data HUb Service (DHuS) - For Space data distribution.
 * Copyright (C) 2013,2014,2015,2016 European Space Agency (ESA)
 * Copyright (C) 2013,2014,2015,2016 GAEL Systems
 * Copyright (C) 2013,2014,2015,2016 Serco Spa
 * 
 * This file is part of DHuS software sources.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
'use strict';

angular
  .module('DHuS-webclient')
.factory('UserService', function($http, $q){
  return{
    model: null,
    signUpUrl:'api/stub/signup',
    userRequestUrl: "api/stub/users/0",       
    signup: function(userModel){
      var self = this;            
      return $http({
            url: ApplicationConfig.baseUrl + self.signUpUrl,
            method: "POST",
            contentType: 'application/json',
            data: JSON.stringify(userModel)
           });
    },
    getUser: function(){
      var self = this;
//      if(self.model){
//        var deferred = $q.defer();
//        deferred.resolve(self.model);
//        return deferred.promise;            
//      }else{
        return $http({url: ApplicationConfig.baseUrl + self.userRequestUrl, method: "GET"})
          .then(function(response) {          
            if (response.status == 200) {
              self.model = response.data;
              return response.data
            }
            else return {};
          });
      //}
    },
    getODataUser: function(username){
      var self = this;
      if(self.model){
        var deferred = $q.defer();
        deferred.resolve(self.model);
        return deferred.promise;            
      }else{
        return $http({url: ApplicationConfig.baseUrl + "odata/v1/Users('" +  username + "')?$format=json", method: "GET"})
          .then(function(response) {          
            if (response.status == 200) {              
              return response.data
            }
            else return {};
          });
      }
    },
    getODataUserRoles: function(username){
      var self = this;
      if(self.model && self.model.roles){
        var deferred = $q.defer();
        deferred.resolve(self.model.roles);
        return deferred.promise;            
      }else{
        return $http({url: ApplicationConfig.baseUrl + "odata/v1/Users('" +  username + "')/SystemRoles?$format=json", method: "GET"})
          .then(function(response) {          
            if (response.status == 200) {              
              return response.data
            }
            else return {};
          });
      }
    },
    /* to be removed after having replaced all stubs related to users with OData BEGIN*/
    setUserModel:function(response){       
      var self=this;
      if(response && response.d) {
        self.model={};
        self.model.username=response.d.Username;
        self.model.password=response.d.Password;
        self.model.firstname=response.d.FirstName;
        self.model.lastname=response.d.LastName;                
        self.model.email=response.d.Email;
        self.model.domain=response.d.Domain;
        self.model.subDomain=response.d.SubDomain;
        self.model.usage=response.d.Usage;
        self.model.subUsage=response.d.SubUsage;
        self.model.country=response.d.Country;       
      }

    },
    setUserRolesModel:function(response){
      var self=this;
      if(response && response.d && response.d.results) {
        if(self.model) {  
          var roles=[];        
          for(var i=0; i < response.d.results.length; i++) {
            roles.push(response.d.results[i].Name);
          }
          self.model.roles=roles;
        }
      }

    },
    /* to be removed after having replaced all stubs related to users with OData END*/
    refreshModel: function(){
      var self=this;
      this.getUser().then(function(user){self.model = user;});
    },
    updateUser: function(user, passwordModel, delegate){
      var self = this;
      self.model = user;
      return $http({
        url: ApplicationConfig.baseUrl + self.userRequestUrl,
        method: "PUT", 
        contentType: 'application/json',         
        data: {
        	user: user,
        	pm: passwordModel
         }
        ,       
        headers: {'Content-Type': 'application/json',
    			  'Accept':'application/json'}        
      });
    },
    retrievePassword: function(user, delegate) {
      var self = this;      
      return $http({
        url: ApplicationConfig.baseUrl + "api/stub/forgotpwd",
        method: "POST", 
        contentType: 'application/json',         
        data: JSON.stringify(user),       
        headers: {'Content-Type': 'application/json',
            'Accept':'application/json'}        
      }).then(function(response) {
          return response.data;
        });
    },
    checkUserCodeForPasswordReset: function(code) {
      var self = this;      
      return $http({
        url: ApplicationConfig.baseUrl + "api/stub/usercode?code="+code,
        method: "GET"               
      });
    },
    resetPassword: function(code, password) {
      var self = this;      
      return $http({
        url: ApplicationConfig.baseUrl + "api/stub/resetpwd?code="
            +code+"&password="+password,
        method: "POST" 
      });
    },    
  };
});



